### PR TITLE
sink: handle row-based changed event in sink, part four

### LINKED ISSUE #272: Sink interface should accept row and resolve event
so can support some sink don't care txn.
the implementation of the sink can reconstruct txn by resolve event internally if need.

### PR DESCRIPTION
**Pull request by leoppro**
### What problem does this PR solve? 
fix #272 , #265 

this pr can pass the integration test now

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-cdc/cdc-ansible`
 - Need to be included in the release note

### CODE REVIEW

File: cdc/entry/schema_builder.go [hunk 1]
```diff
@@ -6,33 +6,58 @@ import (
...
-		currentJob = l.list.Front()
+
+	if ts < l.gcTs {
+		log.Fatal("cannot fetch the jobs which of finishedTs is less then gcTs, please report a bug", zap.Uint64("gcTs", l.gcTs))
 	}
-	for ; currentJob != nil; currentJob = currentJob.Next() {
+
+	if currentJob.Value != (*timodel.Job)(nil) {
 		job := currentJob.Value.(*timodel.Job)
+		if job.BinlogInfo.FinishedTS <= l.gcTs {
```
[the remark this answers is not in the document]
Comment (leoppro) [reply cdc/entry/schema_builder.go]: no, `gcTs` could between` currentJob.FinishedTs` and `nextJob.FinishedTs`

when the `currentJob` is the lastest consumed job, this job maybe dropped already, so we need to check whether the `currentJob` is deleted.

File: cdc/sink/mysql.go [hunk 2]
```diff
@@ -36,37 +37,168 @@ import (
...
+	return atomic.LoadUint64(&s.checkpointTs)
+}
+
+func (s *mysqlSink) Run(ctx context.Context) error {
+	execRows := func(rowsMap map[string][]*model.RowChangedEvent) error {
+		groups := make([][]*model.RowChangedEvent, 0, len(rowsMap))
+		for _, rows := range rowsMap {
+			groups = append(groups, rows)
+		}
+		return s.concurrentExec(ctx, groups)
```
Comment (amyangfei) [inline cdc/sink/mysql.go]: we don't need to convert the `map[string][]*model.RowChangedEvent` to `[][]*model.RowChangedEvent`, we can change the parameter of `concurrentExec` and pass the map to it directly

File: cdc/processor.go [hunk 3]
```diff
@@ -248,7 +177,7 @@ func NewProcessor(
 	}
 
 	for _, table := range p.status.TableInfos {
-		p.addTable(ctx, int64(table.ID), table.StartTs)
+		go p.addTable(ctx, int64(table.ID), table.StartTs)
```
Comment (amyangfei) [inline cdc/processor.go]: why using independent routines here

Comment (leoppro) [reply cdc/processor.go]: A. we create a  processor follow the steps below:
1. create ddl puller
2. create schema builder
3. call `addTable` function
4. call `processor.Run` function, the `Run` of ddl puller will be called in this step. 

B. in `addTable` function
1. call `scheam builder` to create scheam storage, and `schema builder` will call `HandlePreviousDDLJobIfNeed` of new schema storage.
2. do something ...

In step `B.1`, if schema builder is unresolved,  `HandlePreviousDDLJobIfNeed` will return a `ErrUnresolved` error, and schema builder will retry and wait for it.
but ddl puller is still not running, so schema builder will never be resolved.

Comment (amyangfei) [reply cdc/processor.go]: Oh I see. 
`addTable` relies on `ddl puller`

Comment (amyangfei) [reply cdc/processor.go]: The processor startup procedure is not very clear, we can refine this later.

**Pull request merged**